src/scripts/main.js

```diff
@@ -1,3 +1,394 @@
+  } else {
+    restart();
+    updateGameFields();
+    appearingOneRandomCell();
+    updateGameFields();
+    appearingOneRandomCell();
+    updateGameFields();
+  }
+});
+
+function getRandonCell() {
+  const emptyCells = [];
+
+  table.forEach((row, rowIndex) => {
+    row.forEach((cell, columnIndex) => {
+      if (cell === 0) {
+        emptyCells.push([rowIndex, columnIndex]);
+      }
+    });
+  });
+
+  return emptyCells[Math.floor(Math.random() * (emptyCells.length - 1))];
+}
+
+function left(paramArr) {
+  const arr = paramArr.filter(item => item !== 0);
+
+  for (let i = 0; i < arr.length - 1; i++) {
+    const current = arr[i];
+    const next = arr[i + 1];
+
+    if (current === next) {
+      arr[i] = current * 2;
+      scoreValue += arr[i];
+      updateScore();
+      arr[i + 1] = 0;
+    }
+  }
+
+  const mergedArr = arr.filter(item => item !== 0);
+
+  paramArr[0] = mergedArr[0] || 0;
+  paramArr[1] = mergedArr[1] || 0;
+  paramArr[2] = mergedArr[2] || 0;
+  paramArr[3] = mergedArr[3] || 0;
+}
+
+function right(paramArr) {
+  const arr = paramArr.filter(item => item !== 0);
+
+  for (let i = arr.length - 1; i >= 0; i--) {
+    const current = arr[i];
+    const prev = arr[i - 1];
+
+    if (current === prev) {
+      arr[i] = current * 2;
+      scoreValue += arr[i];
+      updateScore();
+      arr[i - 1] = 0;
+    }
+  }
+
+  const mergedArr = arr.filter(item => item !== 0);
+
+  while (mergedArr.length < 4) {
+    mergedArr.unshift(0);
+  }
+
+  paramArr[0] = mergedArr[0] || 0;
+  paramArr[1] = mergedArr[1] || 0;
+  paramArr[2] = mergedArr[2] || 0;
+  paramArr[3] = mergedArr[3] || 0;
+}
+
+function moveHorizontally(directionFunctionName) {
+  const json = JSON.stringify(table);
+  const copyTable = JSON.parse(json);
+
+  for (const row of table) {
+    directionFunctionName(row);
+  }
+
+  for (let i = 0; i < 16; i++) {
+    const rowIndex = Math.floor(i / 4);
+    const colIndex = i - (rowIndex * 4);
+
+    const tableValue = table[rowIndex][colIndex];
+    const copyTableValue = copyTable[rowIndex][colIndex];
+
+    if (tableValue !== copyTableValue) {
+      appearingOneRandomCell();
+      updateGameFields();
+
+      break;
+    }
+  }
+}
+
+function moveVertically(directionFunctionName) {
+  const json = JSON.stringify(table);
+  const copyTable = JSON.parse(json);
+
+  for (let j = 0; j < 4; j++) {
+    directionFunctionName(
+      [table[0][j], table[1][j], table[2][j], table[3][j]], j
+    );
+  }
+
+  for (let i = 0; i < 16; i++) {
+    const rowIndex = Math.floor(i / 4);
+    const colIndex = i - (rowIndex * 4);
+
+    const tableValue = table[rowIndex][colIndex];
+    const copyTableValue = copyTable[rowIndex][colIndex];
+
+    if (tableValue !== copyTableValue) {
+      appearingOneRandomCell();
+      updateGameFields();
+
+      break;
+    }
+  }
+}
+
+document.addEventListener('keydown', eventParam => {
+  if (eventParam.key === 'ArrowLeft') {
+    moveHorizontally(left);
+    loseCheck();
```

Comment:
I'd recommend you to optimize your `loseCheck`. First check if noEmptyCell and only then, if it's true, check noCellToMerge. Since now your start two extra loops inside noCellToMerge each user step

Also, you can invoke this function once after `if` conditions instead of invoking for each if condition